kmath-core/src/commonMain/kotlin/scientifik/kmath/operations/KBigInteger.kt

```diff
@@ -0,0 +1,454 @@
+package scientifik.kmath.operations
+
+import kotlin.math.log2
+import kotlin.math.max
+import kotlin.math.min
+import kotlin.math.sign
+
+/*
+ * Kotlin Multiplatform implementation of Big Integer numbers (KBigInteger).
+ * Initial version from https://github.com/robdrynkin/kotlin-big-integer
+ */
+
+@kotlin.ExperimentalUnsignedTypes
+typealias Magnitude = UIntArray
+
+@kotlin.ExperimentalUnsignedTypes
+typealias TBase = ULong
+
+object KBigIntegerRing: Ring<KBigInteger> {
+    override val zero: KBigInteger = KBigInteger.ZERO
+    override val one: KBigInteger = KBigInteger.ONE
+
+    override fun add(a: KBigInteger, b: KBigInteger): KBigInteger = a.plus(b)
+
+    override fun multiply(a: KBigInteger, k: Number): KBigInteger = a.times(k.toLong())
+
+    override fun multiply(a: KBigInteger, b: KBigInteger): KBigInteger = a.times(b)
+
+    operator fun String.unaryPlus(): KBigInteger = KBigInteger(this)!!
+
+    operator fun String.unaryMinus(): KBigInteger = -KBigInteger(this)!!
+}
+
+
+@kotlin.ExperimentalUnsignedTypes
+class KBigInteger(sign: Int = 0, magnitude: Magnitude = Magnitude(0)):
```

Comment:
Now using Byte for that

Reply:
Even byte is wasteful. Especially considering that my eat whole word due to alignment. Why not use 2-complement?

Reply:
This may be a good idea, but I suspect (performance) problems with 2-complement will appear when we try to operate with numbers of different length. At least the classical Java BigInt implementation stores magnitude and signum separately (http://developer.classpath.org/doc/java/math/BigInteger-source.html)